File: ScapesCore/src/main/java/org/tobi29/scapes/server/ControlPanel.java
package org.tobi29.scapes.server;

import java8.util.Optional;
import java8.util.function.DoubleSupplier;
import java8.util.function.Function;
import java8.util.stream.Collectors;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.tobi29.scapes.engine.server.ControlPanelProtocol;
import org.tobi29.scapes.engine.utils.CPUUtil;
import org.tobi29.scapes.engine.utils.StringUtil;
import org.tobi29.scapes.server.command.Command;
import org.tobi29.scapes.server.connection.PlayerConnection;

import java.util.List;

public class ControlPanel implements Command.Executor {
    private static final Logger LOGGER =
            LoggerFactory.getLogger(ControlPanel.class);
    private static final Function<String, String> ESCAPE = StringUtil
            .replace("&", "&amp;", ">", "&gt;", "<", "&lt;", "'", "&apos;",
                    "\"", "&quot;")::apply;
    private final ControlPanelProtocol connection;
    private final ScapesServer server;

    public ControlPanel(ControlPanelProtocol connection, ScapesServer server) {
        this.connection = connection;
        this.server = server;
        initCommands();
        LOGGER.info("Control panel accepted from {}", connection);
    }

    private void initCommands() {
        connection.addCommand("Ping:Ping", command -> {
            if (command.length >= 1) {
                connection.send("Ping:Pong", command[0]);
            }
        });
        connection.addCommand("Console:Command", command -> {
            if (command.length != 1) {
                return;
            }
            server.commandRegistry().get(command[0], this).execute().forEach(
                    output -> message(output.toString(),
                            MessageLevel.FEEDBACK_ERROR));
        });
        Optional<CPUUtil.Reader> cpuReader = CPUUtil.reader();
        DoubleSupplier cpuSupplier;
        if (cpuReader.isPresent()) {
            CPUUtil.Reader reader = cpuReader.get();
            cpuSupplier = reader::totalCPU;
        } else {
            cpuSupplier = () -> Double.NaN;
        }
        Runtime runtime = Runtime.getRuntime();
        connection.addCommand("Stat:List", command -> {
            double cpu = cpuSupplier.getAsDouble();
            long memory = runtime.totalMemory() - runtime.freeMemory();
            connection.send("Stat:Send", "CPU", String.valueOf(cpu), "Memory",
                    String.valueOf(memory));
        });
        connection.addCommand("Players:List", command -> {
            List<String> players = server.connection().players()
                    .map(PlayerConnection::name)
                    .collect(Collectors.toList());
            connection.send("Players:Send",
                    players.toArray(new String[players.size()]));
        });
    }

    @Override
    public Optional<String> playerName() {
        return Optional.empty();
    }

    @Override
    public String name() {
        return "Control Panel";
    }

    @Override
    public boolean message(String message, MessageLevel level) {
        String style;
        switch (level) {
            case SERVER_ERROR:
            case FEEDBACK_ERROR:
                style = "color: red;";
                break;
            default:
                style = "";
                break;
        }
        message = ESCAPE.apply(message);
        String html = "<span style=\"" + style + "\">" + message + "</span>";
        connection.send("Core:Message", html);
        return true;
    }

    @Override
    public int permissionLevel() {
        return 10;
    }
}

File: ScapesCore/src/main/kotlin/org/tobi29/scapes/server/ControlPanel.kt
/*
 * Copyright 2012-2017 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.server

import org.tobi29.scapes.engine.server.ConnectionWorker
import org.tobi29.scapes.engine.server.ControlPanelProtocol
import org.tobi29.scapes.engine.server.PacketBundleChannel
import org.tobi29.scapes.engine.utils.CPUUtil
import org.tobi29.scapes.engine.utils.ListenerRegistrar
import org.tobi29.scapes.engine.utils.logging.KLogging
import org.tobi29.scapes.engine.utils.tag.*
import org.tobi29.scapes.engine.utils.toReplace
import org.tobi29.scapes.server.command.Executor
import org.tobi29.scapes.server.connection.ServerConnection
import org.tobi29.scapes.server.extension.event.MessageEvent

class ControlPanel(worker: ConnectionWorker,
                   channel: PacketBundleChannel,
                   private val connection: ServerConnection) : ControlPanelProtocol(
        worker, channel, connection.server.events), Executor {

    init {
        initCommands()
    }

    override fun ListenerRegistrar.listeners() {
        listen<MessageEvent> { event ->
            val style: String
            when (event.level) {
                MessageLevel.SERVER_ERROR,
                MessageLevel.FEEDBACK_ERROR -> style = "color: red;"
                else -> style = ""
            }
            val html = "<span style=\"$style\">${ControlPanel.Companion.ESCAPE(
                    event.message)}</span>"
            send("Message", TagMap { this["Message"] = html.toTag() })
        }
    }

    private fun initCommands() {
        addCommand("Ping") { payload ->
            send("Pong", payload)
        }
        addCommand("Command") { payload ->
            payload["Command"]?.toString()?.let { command ->
                connection.server.commandRegistry()[command, this].execute().forEach { output ->
                    events.fire(
                            MessageEvent(this, MessageLevel.FEEDBACK_ERROR,
                                    output.toString(), this))
                }
            }
            payload["Commands"]?.toList()?.asSequence()?.mapNotNull(
                    Tag::toString)?.forEach { command ->
                connection.server.commandRegistry()[command, this].execute().forEach { output ->
                    events.fire(
                            MessageEvent(this, MessageLevel.FEEDBACK_ERROR,
                                    output.toString(), this))
                }
            }
        }
        val cpuReader = CPUUtil.reader()
        val cpuSupplier = cpuReader?.let { { it.totalCPU() } } ?: { Double.NaN }
        val runtime = Runtime.getRuntime()
        addCommand("Stats") {
            val cpu = cpuSupplier()
            val memory = runtime.totalMemory() - runtime.freeMemory()
            send("Stats", TagMap {
                this["CPU"] = cpu.toTag()
                this["Memory"] = memory.toTag()
            })
        }
        addCommand("Players-List") {
            send("Players-List", TagMap {
                this["Players"] = TagList {
                    connection.players.asSequence().map { it.name() }.map {
                        TagMap { this["Name"] = it.toTag() }
                    }.forEach { add(it) }
                }
            })
        }
    }

    override fun playerName(): String? {
        return null
    }

    override fun name(): String {
        return "Control Panel"
    }

    override fun permissionLevel(): Int {
        return 10
    }

    companion object : KLogging() {
        private val ESCAPE = listOf(
                "&" to "&amp",
                ">" to "&gt;",
                "<" to "&lt;",
                "'" to "&apos;",
                "\"" to "&quot;").toReplace()
    }
}
